cmd/syncComparator.go

```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
+			if folderPresent {
+				f.FinalizeTargetDirectory(foldermap, destinationObject.isFinalizeAll)
+			}
+			return nil
+		} else {
+			if folderPresent {
+				sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+				// Lets check if entry in source is folder and entry on destination should be folder.
+				// Otherwise folder in source changed to file.
+				if present && (destinationObject.entityType == sourceObjectInMap.entityType) {
+					delete(foldermap.indexMap, lcFileName)
+					size := storedObjectSize(sourceObjectInMap)
+					size = -size
+					atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+					if sourceObjectInMap.relativePath != destinationObject.relativePath {
+						panic("Relative Path not matched")
+					}
+					delete(f.sourceFolderIndex.folderMap[lcFolderName].indexMap, lcFileName)
+
+					// In case of folder if ctime/mtime changed, lets create the folder.
+					dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+					if f.disableComparison || dataChanged || metadataChanged {
+						err := f.copyTransferScheduler(sourceObjectInMap)
+						if err != nil {
+							return err
+						}
+					}
+					return nil
+				}
+			}
+			// We detect folder not present on source, now we need to delete the folder and files underneath.
+			// Other case will be folder at source changed to file, so we need to delete the folder in target.
+			// TODO: Need to add call to delete the folder.
+			_ = f.destinationCleaner(destinationObject)
+
+			return nil
+		}
+	}
+
+	// File case.
+	// Parent Folder present, we need to check if file exists or not.
+	if folderPresent {
+		sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+
+		// if the destinationObject is present at source and stale, we transfer the up-to-date version from source
+		if present && (sourceObjectInMap.entityType == destinationObject.entityType) {
+			// Sanity check.
+			if sourceObjectInMap.relativePath != destinationObject.relativePath {
+				panic("Relative Path not matched")
+			}
+
+			dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+			if f.disableComparison || dataChanged {
+				err := f.copyTransferScheduler(sourceObjectInMap)
+				if err != nil {
+					return err
+				}
+			} else if metadataChanged {
+				// TODO: Need to add call to just update the metadata only.
+			}
+			size := storedObjectSize(sourceObjectInMap)
+			size = -size
+			atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+			delete(f.sourceFolderIndex.folderMap[lcFolderName].indexMap, lcFileName)
+			return nil
+		}
 	}
 
+	// Parent folder not present this may happen when all the entries for that folder in folder map processed.
+	// In that case the foldemap will be deleted, so we know this file not present in source.
```

Comment:
again, we need this code from doc

                //
                // tobject is a file.
                //          
                // If not present in source, it means the object was deleted
                // from the source since last sync.
                //
                if !IsPresent(ObjectIndexer[dirName][baseName]) {
                    //
                    // Delete the target file unless KeepDeletedFiles is true.
                    //
                    if (!KeepDeletedFiles) {
                       //
	             // For directories DeleteFromTarget will do recursive
	             // deletion.
	             //
	             DeleteFromTarget(tobject)
	          }
	      } else {
	          //
	          // Object with same pathname exists at the source.
	          // Get the storedObject from the ObjectIndexer map,
	          // and delete from ObjectIndexer map.
	          // We delete all processed objects from ObjectIndexer
	          // map so that at the end whatever is left in the map
	          // are new objects created at the source since last
	          // sync and we can then copy them to target.
	          //
	          storedObject = ObjectIndexer[dirName][baseName]
	          delete(ObjectIndexer[dirName][baseName])
	          if (IsFile(storedObject.Attributes)) {
	              //
	              // Target is a regular file and so is the source.
	              //    
	              //
	              // Find out whether file's data or metadata or both
	              // have changed since last sync.
	              // Note that since a "changed directory" can be a renamed
	              // directory, we need to compare each of its children with
	              // their counterpart in the target to find out if they
	              // need to be synced.
	              //
	              dataChanged, metadataChanged =
	           HasFileChangedSinceLastSyncUsingTargetCompare(tobject, storedObject)
	         
	              if dataChanged {
	                  CopyFileToTarget(storedObject)
	              } else if metadataChanged {
	                  CopyFileMetadataToTarget(storedObject)
	              }                    
	          } else if (!IsDirectory(storedObject.Attributes)) {
	              //
	              // Source is a regular file but target is some other
	              // non-directory object like symlink, device file, etc.
	              //
	              DeleteFromTarget(tobject)
	              CopyFileToTarget(storedObject)
	          } else {
	               //
	               // Source is a regular file but target is a directory
	               // (with the same name).
	               //
	               DeleteFromTarget(tobject)
	               CopyFileToTarget(storedObject)
	          }
	      }

let's discuss